File: app/src/main/java/com/jinvoice/data/db/dao/InvoiceDao.kt
package com.jinvoice.data.db.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Transaction
import com.jinvoice.data.db.entity.InvoiceFtsEntity
import com.jinvoice.data.db.entity.InvoiceMetaEntity
import com.jinvoice.data.db.entity.LineItemEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface InvoiceDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertInvoice(invoice: InvoiceMetaEntity): Long

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertLineItems(items: List<LineItemEntity>)

    @Transaction
    suspend fun insertInvoiceWithItems(invoice: InvoiceMetaEntity, items: List<LineItemEntity>) {
        val id = insertInvoice(invoice)
        if (items.isNotEmpty()) {
            insertLineItems(items.map { it.copy(invoiceId = id) })
        }
    }

    /**
     * FTS5 MATCH search — always column-scoped to avoid unscoped full-table FTS.
     * Never use LIKE '%query%' for search.
     */
    @Query("""
        SELECT m.* FROM invoice_meta m
        INNER JOIN invoice_fts f ON m.id = f.rowid
        WHERE invoice_fts MATCH 'merchant_name:' || :query || ' OR merchant_address:' || :query
        ORDER BY m.invoice_date DESC
        LIMIT 50
    """)
    fun searchByMerchant(query: String): Flow<List<InvoiceMetaEntity>>

    @Query("""
        SELECT m.* FROM invoice_meta m
        INNER JOIN invoice_fts f ON m.id = f.rowid
        WHERE invoice_fts MATCH 'merchant_name:' || :query
        ORDER BY m.invoice_date DESC
        LIMIT 50
    """)
    fun searchByMerchantName(query: String): Flow<List<InvoiceMetaEntity>>

    @Query("SELECT * FROM invoice_meta ORDER BY invoice_date DESC LIMIT :limit OFFSET :offset")
    fun getRecentInvoices(limit: Int = 50, offset: Int = 0): Flow<List<InvoiceMetaEntity>>

    @Query("SELECT * FROM invoice_meta WHERE status = 'pending_review' ORDER BY created_at DESC")
    fun getPendingReviewInvoices(): Flow<List<InvoiceMetaEntity>>

    @Query("SELECT * FROM line_item WHERE invoice_id = :invoiceId")
    suspend fun getLineItems(invoiceId: Long): List<LineItemEntity>

    @Query("UPDATE invoice_meta SET status = :status, updated_at = :updatedAt WHERE id = :id")
    suspend fun updateStatus(id: Long, status: String, updatedAt: String)

    @Query("DELETE FROM invoice_meta WHERE id = :id")
    suspend fun deleteInvoice(id: Long)

    /** Rebuild FTS content — called by the repair script after manual index corruption. */
    @Query("INSERT INTO invoice_fts(invoice_fts) VALUES('rebuild')")
    suspend fun rebuildFtsIndex()
}

File: iOS/Sources/Data/DB/InvoiceDAO.swift
import Foundation
import GRDB

final class InvoiceDAO {
    private let db: InvoiceDatabase

    init(db: InvoiceDatabase = .shared) {
        self.db = db
    }

    func insertInvoiceWithItems(invoice: InvoiceMetaRecord, items: [LineItemRecord]) throws {
        try db.dbPool.write { conn in
            var record = invoice
            try record.insert(conn)
            let invoiceId = record.id!
            for var item in items {
                item.invoiceId = invoiceId
                try item.insert(conn)
            }
        }
    }

    func isAlreadyImported(messageId: String) throws -> Bool {
        try db.dbPool.read { conn in
            try ImportRecord.filter(Column("message_id") == messageId).fetchCount(conn) > 0
        }
    }

    func recordImport(_ record: ImportRecord) throws {
        try db.dbPool.write { conn in
            var r = record
            try r.insert(conn)
        }
    }

    func updateStatus(invoiceId: Int64, status: String) throws {
        try db.dbPool.write { conn in
            try conn.execute(
                sql: "UPDATE invoice_meta SET status = ?, updated_at = ? WHERE id = ?",
                arguments: [status, ISO8601DateFormatter().string(from: Date()), invoiceId]
            )
        }
    }

    func searchByMerchant(query: String) throws -> [InvoiceMetaRecord] {
        try db.dbPool.read { conn in
            try InvoiceMetaRecord.fetchAll(conn, sql: """
                SELECT m.* FROM invoice_meta m
                INNER JOIN invoice_fts f ON m.id = f.rowid
                WHERE invoice_fts MATCH 'merchant_name:' || ? || ' OR merchant_address:' || ?
                ORDER BY m.invoice_date DESC LIMIT 50
            """, arguments: [query, query])
        }
    }

    func rebuildFtsIndex() throws {
        try db.dbPool.write { conn in
            try conn.execute(sql: "INSERT INTO invoice_fts(invoice_fts) VALUES ('rebuild')")
        }
    }

    func getPendingReview() throws -> [InvoiceMetaRecord] {
        try db.dbPool.read { conn in
            try InvoiceMetaRecord.filter(Column("status") == "pending_review").fetchAll(conn)
        }
    }
}
